internal/controller/argorollout_controller.go

```diff
@@ -0,0 +1,143 @@
+	"fmt"
+	"net/url"
+	"time"
+
+	clientset "github.com/argoproj/argo-rollouts/pkg/client/clientset/versioned/typed/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/pkg/kubectl-argo-rollouts/cmd/abort"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts"
+	rolloutv1alpha1 "github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	roclientset "github.com/argoproj/argo-rollouts/pkg/client/clientset/versioned"
+	console "github.com/pluralsh/console-client-go"
+	"github.com/pluralsh/deployment-operator/internal/utils"
+	"github.com/pluralsh/deployment-operator/pkg/client"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	k8sClient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	inventoryAnnotationName = "config.k8s.io/owning-inventory"
+	closed                  = "closed"
+)
+
+var requeueRollout = ctrl.Result{RequeueAfter: time.Second * 5}
+
+// ArgoRolloutReconciler reconciles a Argo Rollout custom resource.
+type ArgoRolloutReconciler struct {
+	k8sClient.Client
+	Scheme        *runtime.Scheme
+	ConsoleClient client.Client
+	ConsoleURL    string
+	CachedClient  *utils.CachedClient
+	ArgoClientSet roclientset.Interface
+	DynamicClient dynamic.Interface
+	KubeClient    kubernetes.Interface
+}
+
+// Reconcile Argo Rollout custom resources to ensure that Console stays in sync with Kubernetes cluster.
+func (r *ArgoRolloutReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// Read resource from Kubernetes cluster.
+	rollout := &rolloutv1alpha1.Rollout{}
+	if err := r.Get(ctx, req.NamespacedName, rollout); err != nil {
+		logger.Error(err, "unable to fetch rollout")
+		return ctrl.Result{}, k8sClient.IgnoreNotFound(err)
+	}
+
+	if !rollout.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	serviceID, ok := rollout.Annotations[inventoryAnnotationName]
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+	if serviceID == "" {
+		return ctrl.Result{}, fmt.Errorf("the service ID from the inventory annotation is empty")
+	}
+	service, err := r.ConsoleClient.GetService(serviceID)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	consoleURL, err := sanitizeURL(r.ConsoleURL)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if rollout.Status.Phase == rolloutv1alpha1.RolloutPhasePaused {
+		// wait until the agent will change component status
+		if !hasPausedRolloutComponent(service) {
+			return requeueRollout, nil
+		}
+
+		rolloutIf := r.ArgoClientSet.ArgoprojV1alpha1().Rollouts(rollout.Namespace)
+		promoteURL := fmt.Sprintf("%s/ext/v1/gate/%s", consoleURL, serviceID)
+		rollbackURL := fmt.Sprintf("%s/ext/v1/rollback/%s", consoleURL, serviceID)
+
+		promoteResponse, err := r.CachedClient.Get(promoteURL)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		rollbackResponse, err := r.CachedClient.Get(rollbackURL)
```

Comment:
Caching here might not be best because it'll reduce the responsiveness of the system (might not matter at the end of the day but still should keep it in mind).